contracts/programs/gummyshop/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+        data_hash: [u8; 32],
+        nonce: u128,
+        index: u32,
+        root: [u8; 32],
+        creator_shares: Vec<u8>
+    ) -> Result<()> {
+        // First, payout the marketplace's royalty fee. In this implementation, the marketplace is taking the fee off the top.
+        // However, the marketplace could also take its fee along side the "creators". However, if that approach is taken, just note
+        // that care should be taken that the sum of percentages across creators and marketplace does not exceed 100.
+
+        // @notice: The risk here is that the NFT can be purchased for free (aside from gas cost) if the list price is less than 10,000 lamports.
+        let basis_point_of_price = price/(10000 as u64);
+        let amount_to_pay_marketplace = basis_point_of_price * (ctx.accounts.marketplace_props.share as u64);
+        assert_eq!(ctx.accounts.marketplace_receiver.key(), ctx.accounts.marketplace_props.payment_receiver.key());
+        invoke(
+            &system_instruction::transfer(
+                &ctx.accounts.purchaser.key(),
+                &ctx.accounts.marketplace_receiver.key(),
+                amount_to_pay_marketplace
+            ),
+            &[
+                ctx.accounts.purchaser.to_account_info(),
+                ctx.accounts.marketplace_receiver.to_account_info(),
+                ctx.accounts.system_program.to_account_info()
+            ]
+        )?;
+
+        // Second, payout each "creator". Creators are an immutable set of secondary marketplace sale royalty recipients.
+        // Simultaneously, collect <address, share> pairs to prepare to compute creator_hash
+        let mut total_remaining_price_allocation = price - amount_to_pay_marketplace;
+        let basis_point_of_remaining_price_allocation = total_remaining_price_allocation/(10000 as u64);
+        let mut amount_paid_out_to_creators = 0;
+        let mut creator_data: Vec<Vec<u8>> = Vec::new();
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter();
+        for share in creator_shares.into_iter() {
+            let current_creator_info = next_account_info(remaining_accounts_iter)?;
+            let amount_to_pay_creator = basis_point_of_remaining_price_allocation * ((share as u64) * (100 as u64));
+            invoke(
+                &system_instruction::transfer(
+                    &ctx.accounts.purchaser.key(),
+                    &current_creator_info.key(),
+                    amount_to_pay_creator
+                ),
+                &[
+                    ctx.accounts.purchaser.to_account_info(),
+                    current_creator_info.clone(),
+                    ctx.accounts.system_program.to_account_info()
+                ]
+            )?;
+            amount_paid_out_to_creators += amount_to_pay_creator;
+            creator_data.push([current_creator_info.key().as_ref(), &[share]].concat());
+        }
+        total_remaining_price_allocation -= amount_paid_out_to_creators;
```

Comment:
It's important to check for overflows here. It's unlikely to happen, but always good to do your due diligence